### PR TITLE
Add new RL Colocate Trainer

### PR DESCRIPTION
**Pull request by jayhenry**
## Summary

This PR introduces the `RLColocateTrainer` — a complete RL training loop for the colocated (shared-GPU) setting — along with an evaluation framework and several component-level refinements to support end-to-end GRPO-style reinforcement learning.

### Key Changes

- **`RLColocateTrainer` (new):** A full-featured RL trainer that orchestrates the rollout → train → weight-sync → evaluate loop in a colocated placement group. It integrates `AgentLoopManager` for both training and evaluation data production, supports TensorBoard/JSONL experiment tracking, and handles data preparation (advantage estimation, sequence context construction, logprobs alignment) for GRPO training.
- **`Evaluator` (new):** A lightweight evaluation component (`Evaluator` + `EvaluatorConfig`) that computes metrics (e.g., accuracy) from rollout samples. Evaluation data production is not covered here, and it's handled by `AgentLoopManager`.
- **`_DatasetSampler.__len__`:** Added `__len__` so the evaluator can query total eval sample count for batch size computation.
- **Abstract `Judger` base class:** Extracted a `Judger` ABC from `NativeJudger` so that both `NativeJudger` and `NativeJudgerRouter` share a common interface. Added `RayJudger` / `RayJudgerProxy` type aliases for better IDE support and type hint.
- **`AgentLoopManagerConfig` adjustment:** It has `SamplerConfig`, `AgentLoopConfig`, `ProduceStrategyConfig`  as members now, and internally constructs the agent loop, produce strategy, and sampler. 
- **`RolloutState` serialization fix:** Added a `field_serializer` for `routed_experts` to gracefully skip `ray.ObjectRef` during Pydantic serialization, preventing `PydanticSerializationError`.
- **Design doc update:** Refined `design/component_rl.py` with updated component APIs including `Storage`, `Sampler`, `CheckpointEngine`, `AdvantageEstimator`, and `Packer` abstractions, along with revised usage examples for colocate and disaggregate scenarios.

## Test 
Verified correct reward curve on GSM8K baseline trained on Dense 0.6B, 8B and MoE 30B models.
image

## TODO
- [x] Simplify `Trainer.__init__` code by replacing  `self.build_XXX()` instead of  `XXXConfig.build` (rl02)
- [x] Save train and eval trajectory(rl05)
- [x] save checkpoint (rl06)
- [x] resume
- [x] set deterministic and seed
- [x] add cli (rl01)
- [x] add ReplayBufferCfg
- [ ] Improve RolloutController relative methods such as `check_health`, `pause_generation`, `restart` etc.
- [x] add verl agent loop (rl03)

### CODE REVIEW

File: xtuner/v1/train/rl_colocate_trainer.py [hunk 1]
```diff
@@ -0,0 +1,872 @@
...
+        self.logger = get_logger(log_dir=log_dir, tag="RLTrainer")
+
+        if skip_checkpoint_validation:
+            patch_default_save_plan()
+
+        # steps
+        self._rollout_steps = rollout_steps
+        # self._total_epochs = total_epochs  # TODO
+        self._cur_step = 0
+        self._global_train_step = 1
```
Comment (YanhuiDua) [inline xtuner/v1/train/rl_colocate_trainer.py]: 这里cur_steps, global_train_step定义有些不一致，这次要不统一呢？都是从0开始

File: xtuner/v1/train/rl_colocate_trainer.py [hunk 2]
```diff
@@ -0,0 +1,872 @@
...
+
+    # evaluate config
+    evaluator_config = EvaluatorConfig(
+        enable_evaluate=True,
+        enable_initial_evaluate=False,  # TODO
+        evaluate_step=evaluate_step,  # TODO
+        compute_metric_func=None,
+    )
+    # Finally, build the trainer
+    trainer = RLColocateTrainer(
```
Comment (YanhuiDua) [inline xtuner/v1/train/rl_colocate_trainer.py:864]: 那还需要RLTrainerConfig吗？

File: xtuner/v1/train/rl_colocate_trainer.py [hunk 3]
```diff
@@ -0,0 +1,872 @@
...
+        replay_buffer_config: dict,
+        # agent loop config
+        # agent_loop_config: AgentLoopConfig,
+        # agent loop manager config
+        # produce_strategy_config: ProduceStrategyConfig,
+        agent_loop_manager_cfg: AgentLoopManagerConfig,
+
+        # eval configs
+        eval_agent_loop_manager_cfg: AgentLoopManagerConfig,
+        evaluator_config: EvaluatorConfig,
```
Comment (YanhuiDua) [inline xtuner/v1/train/rl_colocate_trainer.py:167]: 这两个是不是能合并成一个？

enable_evaluate， enable_initial_evaluate， evaluate_step应该是属于Trainer的config
compute_metric_func, eval_sample_ratio, eval_sample_num 应该是属于evaluator的config，当存在多个evaluator的时候应该是evaluator_config + eval_agent_loop_manager_cfg

Comment (YanhuiDua) [reply xtuner/v1/train/rl_colocate_trainer.py:167]: 或者是不是你的想法是EvalAgentLoopManager已经包含了多数据源

File: xtuner/v1/rl/base/agent_loop_manager.py [hunk 4]
```diff
@@ -3,22 +3,35 @@
...
+    sampler_config: SamplerConfig
 
     def build(
         self,
-        agent_loop: AgentLoop,
-        produce_strategy: ProduceStrategy,
-        sampler: Sampler,
+        rollout_controller: RolloutController,
+        judger: Judger,
+        tokenizer: PreTrainedTokenizer | PreTrainedTokenizerFast,
```
Comment (YanhuiDua) [inline xtuner/v1/rl/base/agent_loop_manager.py:29]: 这个改了的话，单测也帮忙改下吧，test_agent_loop.py

File: xtuner/v1/train/rl_colocate_trainer.py [hunk 5]
```diff
@@ -0,0 +1,872 @@
...
+class RLColocateTrainer:
+    _META_PATH = ".xtuner_rl_colocate_trainer"
+    _EXP_TRACKING_PATH = "exp_tracking"
+
+    # 弱化Trainer：Trainer中代码尽量少，尽量用componet来组织代码。
+    # 目标是像torch一样，让用户自己写init 和 train loop，我们只提供组件。
+    def __init__(
+        self,
+        *,
+        resources: AcceleratorResourcesConfig,
```
Comment (YanhuiDua) [inline xtuner/v1/train/rl_colocate_trainer.py:150]: 这里可能还需要传个cpu_resource_config， JudgerRouter.build支持用户传入pg

File: xtuner/v1/train/rl_colocate_trainer.py (hunk 2 above)
Comment (YanhuiDua) [reply xtuner/v1/train/rl_colocate_trainer.py:864]: 可以直接用现在的rl.py的入口吧

File: xtuner/v1/train/rl_colocate_trainer.py (hunk 5 above)
Comment (YanhuiDua) [reply xtuner/v1/train/rl_colocate_trainer.py:150]: @YanhuiDua 跟JudgerRouterConfig一起改

File: xtuner/v1/train/rl_colocate_trainer.py (hunk 2 above)
Comment (YanhuiDua) [reply xtuner/v1/train/rl_colocate_trainer.py:864]: 后续再加

File: xtuner/v1/train/rl_colocate_trainer.py (hunk 3 above)
Comment (jayhenry) [reply xtuner/v1/train/rl_colocate_trainer.py:167]: enable_evaluate， enable_initial_evaluate， evaluate_step 已调整为Trainer属性

File: xtuner/v1/train/rl_colocate_trainer.py (hunk 1 above)
Comment (jayhenry) [reply xtuner/v1/train/rl_colocate_trainer.py]: 已同步添加 _log_mini_batch_metrics 日志
